src/gateway/ws.rs

```diff
@@ -15,12 +15,75 @@ use url::Url;
 use crate::client::bridge::gateway::ChunkGuildFilter;
 use crate::constants::{self, OpCode};
 use crate::gateway::{CurrentPresence, GatewayError};
-use crate::json::{from_str, json, to_string, Value};
+use crate::json::{from_str, to_string};
 use crate::model::event::GatewayEvent;
-use crate::model::gateway::GatewayIntents;
-use crate::model::id::GuildId;
+use crate::model::gateway::{ActivityType, GatewayIntents};
+use crate::model::id::{GuildId, UserId};
 use crate::{Error, Result};
 
+#[derive(Serialize)]
+struct IdentifyProperties {
+    #[serde(rename = "$browser")]
+    browser: &'static str,
+    #[serde(rename = "$device")]
+    device: &'static str,
+    #[serde(rename = "$os")]
+    os: &'static str,
+}
+
+#[derive(Serialize)]
+struct ActivityData<'a> {
+    #[serde(rename = "type")]
+    kind: ActivityType,
+    url: &'a Option<Url>,
+    name: &'a str,
+}
+
+#[derive(Serialize)]
+struct ChunkGuildMessage<'a> {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    user_ids: Option<Vec<UserId>>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    query: Option<&'a str>,
+    guild_id: GuildId,
+    nonce: &'a str,
+    limit: u16,
+}
+
+#[derive(Serialize)]
+#[serde(untagged)]
+enum WebSocketMessageData<'a> {
+    Heartbeat(Option<u64>),
+    ChunkGuild(ChunkGuildMessage<'a>),
+    Identify {
+        v: u8,
+        compress: bool,
+        shard: &'a [u64; 2],
+        token: &'a str,
+        large_threshold: u8,
+        intents: GatewayIntents,
+        properties: IdentifyProperties,
+    },
+    PresenceUpdate {
+        afk: bool,
+        status: &'a str,
+        since: SystemTime,
+        game: Option<ActivityData<'a>>,
+    },
+    Resume {
+        session_id: &'a str,
+        token: &'a str,
+        seq: u64,
+    },
+}
+
+#[derive(Serialize)]
+struct WebSocketMessage<'a> {
+    op: OpCode,
+    #[serde(borrow)]
```

Comment:
From looking at the serde docs it seems like `serde(borrow)` is used for deserialization only, which doesn't apply here (correct me if i'm wrong)